src/status_im2/contexts/activity_center/events.cljs

```diff
@@ -48,127 +49,59 @@
 
 ;;;; Notification reconciliation
 
-(defn- notification-type->filter-type
-  [type]
-  (if (some types/membership [type])
-    types/membership
-    type))
-
 (defn- update-notifications
-  "Insert `new-notifications` in `db-notifications`.
-
-  Although correct, this is a naive implementation for reconciling notifications
-  because for every notification in `new-notifications`, linear scans will be
-  performed to remove it and sorting will be performed for every new insertion.
-  If the number of existing notifications cached in the app db becomes
-  ~excessively~ big, this implementation will probably need to be revisited."
-  [db-notifications new-notifications]
-  (reduce (fn [acc {:keys [id read] :as notification}]
-            (let [filter-type         (notification-type->filter-type (:type notification))
-                  remove-notification (fn [data]
-                                        (remove #(= id (:id %)) data))
-                  insert-and-sort     (fn [data]
-                                        (->> notification
-                                             (conj data)
-                                             (sort-by (juxt :timestamp :id))
-                                             reverse))]
-              (as-> acc $
-                (update-in $ [filter-type :all :data] remove-notification)
-                (update-in $ [types/no-type :all :data] remove-notification)
-                (update-in $ [filter-type :unread :data] remove-notification)
-                (update-in $ [types/no-type :unread :data] remove-notification)
-                (if (:deleted notification)
-                  $
-                  (cond-> (-> $
-                              (update-in [filter-type :all :data] insert-and-sort)
-                              (update-in [types/no-type :all :data] insert-and-sort))
-                    (not read) (update-in [filter-type :unread :data] insert-and-sort)
-                    (not read) (update-in [types/no-type :unread :data] insert-and-sort))))))
-          db-notifications
-          new-notifications))
+  [db-notifications new-notifications {filter-type :type filter-status :status}]
+  (->> new-notifications
+       (reduce (fn [acc {:keys [id type] :as notification}]
+                 (if (or (:deleted notification)
+                         (and (= :unread filter-status)
+                              (:read notification))
+                         (and (set? filter-type)
+                              (not (contains? filter-type type)))
+                         (and (not (set? filter-type))
+                              (not= filter-type types/no-type)
+                              (not= filter-type type)))
+                   (dissoc acc id)
+                   (assoc acc id notification)))
+               (collection/index-by :id db-notifications))
+       (vals)
+       (sort-by (juxt :timestamp :id))
+       (reverse)))
```

Comment:
could we sort in opposite direction? so reverse is not needed

Reply:
The original implementation did exactly that, but it had a bug caught by the tests. I attempted to quickly understand why and gave up for the moment, but I can take another look.

In theory this is not gonna be a big deal most of the time, because notifications are going to be re-fetched on tab switches and when opening the AC. But, if the user has hundreds of notifications and scroll to the bottom, then when a new one arrives the reconciliation will be a bit slower, but I suspect this is not gonna be the case for most user flows.

IMO, the ideal solution involves NOT storing in the app db all the notifications as the user scrolls and new pages arrive. We should have a cap on the max number of notifications to store in the app db, e.g. using some kind of circular list data structure and dropping the "old" ones as the user scrolls. I'm interested in taking a look at this because it's a problem that I've optimized in the past, but for the web platform.

Reply:
For instance, without a max limit of notifications to store in the app db, if the user leaves the screen turned on in the AC and new signals arrive constantly (imagine you're famous and receives a lot of notifications) the app will eventually start to perform terribly and consume more battery doing all the synchronization logic.

Reply:
For the record: How do we reverse a comparison using `juxt` as the key fn? It wasn't obvious to me. `sort-by` uses the `compare` function to compare the value returned by `(juxt :timestamp :id)`. For example, it would compare `[100 "0x1"]` and `[100 "0x2"]` and return `-1`, because "0x1" is *less than* "0x2". To reverse, we need to pass a comparator function that swaps the arguments passed to `compare`.

Sometimes we just forget about old tricks ;)

```clojure
;; Correctly reverses order
(let [coll [{:id "0x1" :timestamp 1}
            {:id "0x2" :timestamp 50}
            {:id "0x3" :timestamp 50}
            {:id "0x4" :timestamp 100}]]
  (sort-by (juxt :timestamp :id)
           #(compare %2 %1)
           coll))
;; => [{:id "0x4", :timestamp 100}
;;     {:id "0x3", :timestamp 50}
;;     {:id "0x2", :timestamp 50}
;;     {:id "0x1", :timestamp 1}]

```

Reply:
Fixed https://github.com/status-im/status-mobile/pull/15216/commits/394184eff49519eb73248ca75835a38dabf8d58a